### PR TITLE
feat: eth: parse revert data

### PR DESCRIPTION
**Pull request by Stebalien**
## Proposed Changes

Parse the revert data and include it in the error message of EthCall to improve degradability. We don't really want to do this in the FVM because it's Ethereum (solidity) specific, but this makes sense to do in the Ethereum API. It might also make sense to do in the EVM itself, but it's too late to consider that right now.

This change only applies to `EthCall`.  The next step is to improve the error message output from `EthEstimateGas`.

## Additional Info

See https://docs.soliditylang.org/en/latest/control-structures.html#panic-via-assert-and-error-via-require

## Checklist

Before you mark the PR ready for review, please make sure that:

- [x] Commits have a clear commit message.
- [x] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [x] Tests exist for new functionality or change in behavior
- [x] CI is green

### CODE REVIEW

File: node/impl/full/eth.go [hunk 1]
```diff
@@ -363,6 +363,19 @@ func (a *EthModule) EthGetTransactionCount(ctx context.Context, sender ethtypes.
...
+	if err != nil {
+		if xerrors.Is(err, types.ErrActorNotFound) {
+			return 0, nil
+		}
+		return 0, xerrors.Errorf("failed to lookup contract %s: %w", sender, err)
+	}
+
+	switch {
+	case builtinactors.IsAccountActor(actor.Code), builtinactors.IsEthAccountActor(actor.Code):
+
```
Comment (raulk) [inline node/impl/full/eth.go:376]: Looks like leftover code?

Comment (Stebalien) [reply node/impl/full/eth.go:376]: Uh, yes.

File: node/impl/full/eth.go [hunk 2]
```diff
@@ -2180,6 +2181,92 @@ func parseEthTopics(topics ethtypes.EthTopicSpec) (map[string][][]byte, error) {
...
+// See https://docs.soliditylang.org/en/latest/control-structures.html#panic-via-assert-and-error-via-require
+func parseEthRevert(ret []byte) string {
+	if len(ret) == 0 {
+		return "none"
+	}
+	var cbytes abi.CborBytes
+	if err := cbytes.UnmarshalCBOR(bytes.NewReader(ret)); err != nil {
+		return "ERROR: revert reason is not cbor encoded bytes"
+	}
+	if len(cbytes) == 0 {
```
Comment (shrenujbansal) [inline node/impl/full/eth.go:2199]: In what scenario would you run into this and not 2192?

File: node/impl/full/eth.go [hunk 3]
```diff
@@ -2180,6 +2181,92 @@ func parseEthTopics(topics ethtypes.EthTopicSpec) (map[string][][]byte, error) {
...
+			codeInt := big.PositiveFromUnsignedBytes(cbytes)
+			return fmt.Sprintf("Panic(%s)", ethtypes.EthBigInt(codeInt).String())
+		}
+		switch code {
+		case 0x00:
+			return "Panic()"
+		case 0x01:
+			return "Assert()"
+		case 0x11:
+			return "ArithmaticOverflow()"
```
Comment (shrenujbansal) [inline node/impl/full/eth.go:2222]: *Arithmetic

File: node/impl/full/eth.go (hunk 2 above)
Comment (Stebalien) [reply node/impl/full/eth.go:2199]: 2192 means that the contract returned nothing. In this case, the contract returned an empty cbor byte string.

Really, we should _only_ see the second case (right now). We should optimize contracts to return "nothing" in this case, but we haven't.

Comment (Stebalien) [reply node/impl/full/eth.go:2199]: I.e. `void != []`

File: node/impl/full/eth.go [hunk 4]
```diff
@@ -2180,6 +2181,92 @@ func parseEthTopics(topics ethtypes.EthTopicSpec) (map[string][][]byte, error) {
...
+	case panicFunctionSelector:
+		cbytes := cbytes[4 : 4+32]
+		// Read the and check the code.
+		code, err := ethtypes.EthUint64FromBytes(cbytes)
+		if err != nil {
+			// If it's too big, just return the raw value.
+			codeInt := big.PositiveFromUnsignedBytes(cbytes)
+			return fmt.Sprintf("Panic(%s)", ethtypes.EthBigInt(codeInt).String())
+		}
+		switch code {
```
Comment (shrenujbansal) [inline node/impl/full/eth.go:2216]: Where are the below codes defined?
Stick these in a const map and return that?

Comment (Stebalien) [reply node/impl/full/eth.go:2216]: Link above, but you're right. We can just use a map. I'll make one.

File: node/impl/full/eth.go (hunk 2 above)
Comment (shrenujbansal) [reply node/impl/full/eth.go:2199]: ok, so something like
`revert() vs revert("")`

File: itests/fevm_test.go [hunk 5]
```diff
@@ -868,3 +868,32 @@ func TestFEVMGetBlockDifficulty(t *testing.T) {
...
+
+	_, contractAddr := e.DeployContractFromFilename(ctx, "contracts/Errors.hex")
+	contractAddrEth, err := ethtypes.EthAddressFromFilecoinAddress(contractAddr)
+	require.NoError(t, err)
+	for sig, expected := range map[string]string{
+		"failRevertEmpty()":  "none",
+		"failRevertReason()": "Error(my reason)",
+		"failAssert()":       "Assert()",
+		"failDivZero()":      "DivideByZero()",
+		"failCustom()":       "0x09caebf3",
```
Comment (shrenujbansal) [inline itests/fevm_test.go:886]: How do you get "0x09caebf3"?
maybe add a comment on the side to explain

**Review by shrenujbansal — APPROVED** [review]

File: itests/fevm_test.go (hunk 5 above)
Comment (Stebalien) [reply itests/fevm_test.go]: I'm now calculating it inline.

**Pull request merged**